lib/hdkeys.c

```diff
@@ -0,0 +1,257 @@
+		memset(ek->chaincode.data, 0, sizeof(ek->chaincode.data));
+		ek->index = 0;
+		ek->version = 0;
+		memset(ek->parent_fingerprint, 0, 4);
+		ek->depth = 0;
+		return true;
+	}
+	return false;
+}
+
+void hd_extended_key_free(struct hd_extended_key *ek)
+{
+	bp_key_free(&ek->key);
+}
+
+bool hd_extended_key_deser(struct hd_extended_key *ek, const void *_data,
+			   size_t len)
+{
+	if (78 != len && 82 != len) return false;
+
+	struct const_buffer buf = { _data, len };
+	uint32_t version;
+
+	if (!deser_u32(&version, &buf)) return false;
+	ek->version = version = be32toh(version);
+	if (!deser_bytes(&ek->depth, &buf, 1)) return false;
+	if (!deser_bytes(&ek->parent_fingerprint, &buf, 4)) return false;
+	if (!deser_u32(&ek->index, &buf)) return false;
+	ek->index = be32toh(ek->index);
+	if (!deser_bytes(&ek->chaincode.data, &buf, 32)) return false;
+
+	if (MAIN_PUBLIC == version || TEST_PUBLIC == version) {
+		if (bp_pubkey_set(&ek->key, buf.p, 33)) {
+			return true;
+		}
+	} else if (MAIN_PRIVATE == version || TEST_PRIVATE == version) {
+		uint8_t zero;
+		if (deser_bytes(&zero, &buf, 1) && 0 == zero) {
+			if (bp_key_secret_set(&ek->key, buf.p, 32)) {
+				return true;
+			}
+		}
+	}
+
+	return false;
+}
+
+static void hd_extended_key_ser_base(const struct hd_extended_key *ek,
+				     cstring *s, uint32_t version)
+{
+	ser_u32(s, htobe32(version));
+	ser_bytes(s, &ek->depth, 1);
+	ser_bytes(s, &ek->parent_fingerprint, 4);
+	ser_u32(s, htobe32(ek->index));
+	ser_bytes(s, &ek->chaincode, 32);
+}
+
+bool hd_extended_key_ser_pub(const struct hd_extended_key *ek, cstring *s)
+{
+	hd_extended_key_ser_base(ek, s, MAIN_PUBLIC);
+
+	void *pub;
+	size_t pub_len;
+	if (bp_pubkey_get(&ek->key, &pub, &pub_len) && 33 == pub_len) {
+		ser_bytes(s, pub, 33);
+		free(pub);
+		return true;
+	}
+	return false;
+}
+
+bool hd_extended_key_ser_priv(const struct hd_extended_key *ek, cstring *s)
+{
+	hd_extended_key_ser_base(ek, s, MAIN_PRIVATE);
+
+	const uint8_t zero = 0;
+	ser_bytes(s, &zero, 1);
+	return bp_key_secret_get(s->str + s->len, 32, &ek->key);
+}
+
+bool hd_extended_key_generate_master(struct hd_extended_key *ek,
+				     const void *seed, size_t seed_len)
+{
+	static const uint8_t key[12] = "Bitcoin seed";
+	uint8_t I[64];
+	HMAC(EVP_sha512(), key, (int)sizeof(key), (const uint8_t *)seed,
+	     (uint32_t)seed_len, &I[0], NULL);
+
+	if (bp_key_secret_set(&ek->key, I, 32)) {
+		memcpy(ek->chaincode.data, &I[32], 32);
+		ek->index = 0;
+		ek->version = MAIN_PRIVATE; // get's set public / private during
+		memset(ek->parent_fingerprint, 0, 4);
+		ek->depth = 0;
+
+		return true;
+	}
+
+	return false;
+}
+
+bool hd_extended_key_generate_child(const struct hd_extended_key *parent,
+				    uint32_t index,
+				    struct hd_extended_key *out_child)
+{
+	bool result = false;
+
+	// Get parent public key
+
+	void *parent_pub = NULL;
+	size_t parent_pub_len = 0;
+	if (!bp_pubkey_get(&parent->key, &parent_pub, &parent_pub_len)) {
```

Comment:
I believe it works/fails as appropriate wrt hardened keys.  I always extract the parent public key data here because it's needed for the fingerprint (arguably that should be done after the hardened key check below, although the control flow gets messy).

So, if 'parent' is an extended public key then the function behaves like CKDpub from the spec and returns an extended public key.  If 'parent' is an extended private key, it's like CKDpriv and returns a (potentially hardened) extended private key, which can then be serialized to either an extended public or private key as required.

Let me know if I missed something in the above logic and I'll add a test for the case that would fail.